### PR TITLE
3118 central server config through m supply central

### LINKED ISSUE #3118: Configure omSupply central credentials in 4d mSupply
This is to deal with 'is_central_server' flag.

From this comment https://github.com/msupply-foundation/open-msupply/pull/2819#issuecomment-1955364839

* In legacy central mSupply central server configurations with configure which site is 'central' and how it can be reached (this is just stored against the 'sites' table).
* Before every sync operation, there is a call to some 'info' end point, this will double check that it's ok to sync this site and also will return back is this site central omSupply server and where is central omSupply server. That should give us enough flexibility and an easy way to inform existing sites that are updated, where to find omSupply server

### PR DESCRIPTION
**Pull request by andreievg**
Fixes #3118

**IMPORTANT**, use [omSupply-central-server-configurations ](https://github.com/msupply-foundation/msupply/pull/14581) branch of mSupply (will update in a sec and PR to follow)

# 👩🏻‍💻 What does this PR do? 
 
* Adds global mutex containing central server configurations, please have a look at server/service/src/sync/mod.rs first
* Site info has a couple of new fields, that tell all omSupply sites if they are central server, and where it can be reached
* During sync we first call site info to figure out if current site is central and what central site url is
* Fix integration tests
* Update documentation, to explain how omSupply central site can be launched, might need to update more internal docs (to explain how omSupply central URL is exposed in omSupply remote sites, i.e. through `site_info` end point)

# 🧪 How has/should this change been tested? 

## 💌 Any notes for the reviewer?

I had a look at sync version, it's on https://github.com/msupply-foundation/msupply/blob/6609e8078cb5b88bf7086edf8128ca1c7cd88caa/Project/Sources/Methods/syncV5API.4dm#L17C1-L17C16 atm for mSupply. If mSupply changes for this work go into next mSupply release (where iMaxVersion is 4), we should be fine with compatibility, otherwise we should increment to 5 in both (omSupl and mSup)

## 📃 Documentation

_No user facing changes_

or

- [ ] Functional change 1
- [ ] Functional change 2

### CODE REVIEW

File: server/README.md
Comment (mark-prins) [inline server/README.md:5]: wondering about this..?
```suggestion
- Hosts the server web interface and exposes RESTful and GraphQL APIs for open mSupply data.
```

and the list would read more naturally with an intro such as:

```
Open mSupply server is a component of the Open mSupply system. The server:
```

File: server/README.md
Comment (mark-prins) [inline server/README.md:8]:
```suggestion
- Can act as a remote or central omSupply server, see [Synchronisation documentation](https://github.com/msupply-foundation/open-msupply/blob/develop/server/service/src/sync/README.md#glossary)
```

and perhaps 'omSupply server' should be 'Open mSupply server'?

File: server/README.md
Comment (mark-prins) [inline server/README.md:106]: again - probably fine here, in internal docs: the preference for external documentation is to use 'Open mSupply' in preference to 'omSupply'

File: server/README.md
Comment (mark-prins) [inline server/README.md:108]:
```suggestion
- Configure one site to be central server by checking `Site is open mSupply central server` and entering the URL with which remote sites can reach this site in the field `This site url`
```

File: server/service/src/sync/README.md
Comment (mark-prins) [inline server/service/src/sync/README.md:99]:
```suggestion
An instance of omSupply can be configured to identify as omSupply central server by updating the site configuration in legacy mSupply. Special -> Synchronisation -> {site}, `Site is open mSupply central server` checked and `This site url` the url with which other remote sites can reach omSupply central server
```

File: server/service/src/sync/test/integration/README.md
Comment (mark-prins) [inline server/service/src/sync/test/integration/README.md:55]: line 71 - there's a typo of 'centarl' which should be 'central'

**Review by mark-prins — APPROVED** [review]
👌 LGTM - I've suggested some wording changes / spelling nits in the readme. apart from that, the RwLock is neat, and the sync is working nicely. Have tested running the central & remote sites manually and integration tests

File: server/README.md
Comment (andreievg) [reply server/README.md:5]: fdd9d526e72d661880053dabe425bcfae0c3dbd5 🙏

File: server/README.md
Comment (andreievg) [reply server/README.md:106]: Thanks, i'll try to change to use Open mSupply in internal docs too (and in some places I have), it's just so much easier to type omSupply

**Comment by andreievg** [comment]
Thanks for review @mark-prins and all the suggestions. I'll wait for og work to be merged at the same time

Also there is file sync work that looks up omSupply central server by + 2 port offset, which would be merged first, and then i'll update that logic after merging from dev (to use the new way of finding omSupply central server)

**Comment by andreievg** [comment]
I've merged https://github.com/msupply-foundation/open-msupply/pull/3653 into this one and fixed conflict + change file sync to use new configuration (seems to work)

Re-tested with adjusted instructions from https://github.com/msupply-foundation/open-msupply/pull/3653::

You will need to run omSupply central and remote server

Start remote server and front end with:

yarn && yarn start-local -- -- --port 3005
cargo run

front end available on http://localhost:3005/

Start omSupply central with

APP_SERVER__PORT=2055 APP_DATABASE__DATABASE_NAME=central APP_SERVER__BASE_DIR=central_app_data cargo run
yarn start-local -- -- --env API_HOST=http://localhost:2055

front end available on http://localhost:3003/

Initialise both sites (separate sites), make both stores use vaccine module, and the central site would need to be configured as central server with open msupply url: http://localhost:2055/

Add new asset on remote site, add document.
Sync this site and wait for central to receive the file
Check file is available on central

Check that file is present in central_app_data directory under static_files

Stop remote site, and wipe it's database and remove app_data directory.

Re-initialise remote site

Try to view asset document

app_data folder should now have the document

**Pull request merged**
